File: dynamic.js
function fib(num) {
  if (num === 1 || num === 2) {
    return 1
  } else {
    return fib(num - 1) + fib(num - 2)
  }
}

// console.log(fib(7));

function fibDynamic(num, memo) {
  console.log("recursion");
  if (num === 1 || num === 2) {
    return 1
  } else if (num in memo) {
    return memo[num]
  } else {
    memo[num] = fibDynamic(num - 1, memo) + fibDynamic(num - 2, memo)
    return memo[num]
  }
}

// console.log(fibDynamic(7, {}));

// EXERCISE 1
// store `add_until_100(array[1, array.length - 1])` in a variable before the first
// if statement

function golomb(n, memo) {
  if (n === 1) {
    return 1
  } else {
    if (!memo[n]) {
      memo[n] = 1 + golomb(n - golomb(golomb(n - 1, memo), memo), memo);
    }
    return memo[n];
  }
}

//console.log(golomb(5, {}))

function uniquePaths(rows, columns, memo) {
  if (rows === 1 || columns === 1) {
    return 1;
  }
  if (!memo[`${rows - 1}/${columns}`] && !memo[`${rows}/${columns - 1}`]) {
    memo[`${rows - 1}/${columns}`] = uniquePaths(rows - 1, columns, memo)
    memo[`${rows}/${columns - 1}`] = uniquePaths(columns - 1, rows, memo)
  }
  return memo[`${rows - 1}/${columns}`] + memo[`${rows}/${columns - 1}`]
}

console.log(uniquePaths(100, 2, {}))

File: dynamic.go
package main

import "fmt"

func fibDynamic (fibNum int, memo map[int]int) int {
  fmt.Println("RECURSION")
  if fibNum == 0 || fibNum == 1 {
    return fibNum
  } else {
    _, ok := memo[fibNum]
    if !ok {
      memo[fibNum] = fibDynamic(fibNum - 1, memo) + fibDynamic(fibNum - 2, memo)
    }
    return memo[fibNum]
  }
}

func golomb(n int, memo map[int]int) int {
  if n == 1 {
    return 1
  } else {
    _, ok := memo[n]
    if !ok {
      memo[n] = 1 + golomb(n - golomb(golomb(n - 1, memo), memo), memo);
    }
    return memo[n]
  }
}

func uniquePaths(rows int, columns int, memo map[string]int) int {
  if rows == 1 || columns == 1 {
    return 1
  }
  _, ok := memo[fmt.Sprintf("%d, %d", rows, columns)]
  if !ok {
    memo[fmt.Sprintf("%d, %d", rows, columns)] = uniquePaths(rows - 1, columns, memo) + uniquePaths(rows, columns - 1, memo)
  } 
  return memo[fmt.Sprintf("%d, %d", rows, columns)]
}

func main () {
  // fmt.Println(fibDynamic(5, map[int]int{}))
  // fmt.Println(fibDynamic(7, map[int]int{}))
  // fmt.Println(golomb(5, map[int]int{}))
  fmt.Println(uniquePaths(100, 2, map[string]int{}))
}